Dev

Jerry987123 opened the pull request.

Jerry987123 committed:
加上AFNetworking、JSONModel、SDWebImage library

Jerry987123 committed:
實作TabBarController

Jerry987123 committed:
TabBarViewController樣式設定

Jerry987123 committed:
加上SearchController元件

Jerry987123 committed:
建立Movie Cell

Jerry987123 committed:
建立MusicCell

Jerry987123 committed:
裝rxSwift library

Jerry987123 committed:
SearchingView實作Movie和MusicModel串接TableView

Jerry987123 committed:
電影和音樂資料結構連接畫面

Jerry987123 committed:
unitTest測試文字

Jerry987123 committed:
設定Target inherits和開發者帳號

Jerry987123 committed:
降deployment target = 13.0

Jerry987123 committed:
降deployment target = 11.2、刪除SceneDelegate、新增iTunesSearchAPI呼叫

Jerry987123 committed:
串Objective-C bridging header、實作AFNetworking和JSONModel library

Jerry987123 committed:
iTunesSearchAPI response以JSONModel串接

Jerry987123 committed:
調整SearchingAPI呼叫方式（使用Swift包裝Object C）

Commit:
修正錯誤的檔案路徑

Commit:
檔案名稱調整

Commit:
searching movie datas串接tableView資料源

Commit:
Searching串music資料

Commit:
searchbar串接API

Jerry987123 committed:
Launch畫面設定

Jerry987123 committed:
修正未搜尋前有一筆空結果

Jerry987123 committed:
設定readmore按鈕展開和tableview cell版面調整設定

Jerry987123 committed:
實作SDWebImage下載圖片

Jerry987123 committed:
點選搜尋結果開啟 trackViewUrl

Jerry987123 committed:
呼叫 API搜尋時，顯示 loading

Jerry987123 committed:
對搜尋關鍵字做url-encoded

Jerry987123 committed:
實作個人資料->主題顏色->主題顏色設定頁面

Jerry987123 committed:
主題設定實作

Jerry987123 committed:
建立主題色表和切換功能

Jerry987123 committed:
主題顏色調整

Jerry987123 committed:
建立個人頁收藏按鈕和關於按鈕

Jerry987123 committed:
實作webView

Commit:
檔案路徑修正

Commit:
主題切換回到前一頁

Jerry987123 committed:
建立收藏頁面

Jerry987123 committed:
改topic色表

Jerry987123 committed:
收藏畫面版面製作

Jerry987123 committed:
收藏畫面調整

Jerry987123 committed:
點擊收藏轉導至iTunes

Jerry987123 committed:
print db path

Jerry987123 committed:
實作dequeueReusableCell

Jerry987123 committed:
移除未使用的檔案和程式

Jerry987123 committed:
消除ObjectC func命名警告

Jerry987123 committed:
musicCell版面調整、setCell給預設圖片、keyword urlencoded換位置

Jerry987123 committed:
當離開searchView時關閉searchbar

Jerry987123 committed:
變數調整

Jerry987123 committed:
加入DB套件：FMDB、建立收藏表格、CRUD語法、對應的Unit test

Jerry987123 committed:
db model改用iTunesSearchAPIResponseResult、收藏結果頁串接DB

Jerry987123 committed:
insert改成可新增複數資料

Jerry987123 committed:
DB table加上mediaType以區分music和movie、收藏畫面調整

Jerry987123 committed:
移除未使用的code

Jerry987123 committed:
移除未使用的code

Jerry987123 committed:
收藏頁註冊MVVM

Jerry987123 committed:
實作收藏、取消收藏按鈕

Jerry987123 committed:
改Collection table的primary key為trackId

Jerry987123 committed:
顯示收藏數量

Jerry987123 committed:
千字位顯示

Jerry987123 committed:
音樂Cell版面調整

Jerry987123 committed:
搜尋頁的收合動作修正

Jerry987123 committed:
movie cell 收合邏輯調整

Jerry987123 committed:
實作DispatchGroup做2個非同步執行緒

Jerry987123 committed:
主題切換保留在同一頁

Jerry987123 committed:
API錯誤回傳error

Jerry987123 committed:
API網址調回正常

Jerry987123 committed:
RxSwift錯誤處理實作

Jerry987123 committed:
API錯誤處理跳出Alert

Jerry987123 committed:
收藏按鈕依資料庫做名稱調整

Jerry987123 committed:
音樂收藏按鈕名稱顯示邏輯

Jerry987123 committed:
收藏頁電影cell收合功能實作

Jerry987123 committed:
實作收藏電影頁取消收藏

Jerry987123 committed:
收藏音樂頁取消收藏實作

Jerry987123 committed:
回到搜尋頁即更新收藏資訊

Jerry987123 committed:
移除 UITest 的Podfile設定

Jerry987123 committed:
table cell收藏按鈕畫面調整

Jerry987123 committed:
資料筆數調整sql語法

Jerry987123 committed:
測試

Jerry987123 committed:
刪除測試文字

Jerry987123 committed:
主題切換時不會清完搜尋頁結果

Jerry987123 committed:
development team設定

Jerry987123 committed:
修正在收藏頁取消收藏時bug、修正在移除收藏後收合錯誤顯示問題

Jerry987123 committed:
以enum Result處理Observer的收到錯誤的情境

Jerry987123 committed:
itunes API 加入參數country=tw

Jerry987123 committed:
加上weak修飾

markchangjz commented on BootcampTraining/SearchingView/SearchingViewModel/API/iTunesSearchAPIObj.m:
```diff
@@ -0,0 +1,40 @@
+            if (jsonResponseResult != nil) {
+                [results addObject:jsonResponseResult];
+            }
+        }
```
這段可以用 jsonmodel library 中的 [model-collections](https://github.com/jsonmodel/jsonmodel#model-collections) 修正，就可以不用自己寫 for 迴圈

markchangjz replied:
這段可以用 jsonmodel library 中的 [model-collections](https://github.com/jsonmodel/jsonmodel#model-collections) 修正，就可以不用自己寫 for 迴圈

markchangjz commented on BootcampTraining/CollectionView/CollectionViewController.swift:
```diff
@@ -0,0 +1,82 @@
+//
```
* [crash] 點選收藏的音樂 crash

Jerry987123 replied:
已修正

markchangjz replied:
ok

markchangjz commented on BootcampTraining/CollectionView/CollectionViewController.swift:
```diff
@@ -0,0 +1,82 @@
+//
+//  CollectionViewController.swift
```
* [bug] 收藏電影收合狀態錯亂
0. 清空所有收藏項目
1. 搜尋 appple 收藏前兩部電影
2. 至電影收藏頁，點選第一筆項目 "read more" 按鈕，狀態會變成 'read less' 按鈕
3. 回到搜尋頁，取消第二筆收藏
4. 再回到電影收藏頁，已展開的項目按鈕變成 'read more'，預期仍是 'read less' 按鈕

![pic1](https://user-images.githubusercontent.com/15119448/124212178-15bc5580-db21-11eb-9af5-73c8c61984bd.jpeg)

Jerry987123 replied:
已修正

markchangjz replied:
ok

markchangjz replied:
ok

Jerry987123 committed:
點選收藏的音樂 crash 修正

markchangjz commented on BootcampTraining/SearchingView/SearchingViewModel/API/iTunesSearchAPIObj.m:
```diff
@@ -0,0 +1,40 @@
+
+        // Call completion handler.
+        self->_completionHandler(results);
+
```
印象中有 ARC 後，就可以不用再 `[handler copy];`，

直接寫 `handler(results);` 即可，

`void (^_completionHandler)(NSMutableArray<iTunesSearchAPIResponseResult*> *someParameter);` 這個變數就可以拿掉了

Jerry987123 replied:
已修正

markchangjz replied:
ok

Jerry987123 committed:
拿掉_completionHandler

Jerry987123 committed:
移除self.前綴、移除未使用程式碼

markchangjz commented on BootcampTraining/TopicView/TopicViewController.swift:
```diff
@@ -0,0 +1,75 @@
+        navigationItem.backBarButtonItem?.tintColor = .black
+    }
+    private func setTableView(){
+        if _tableView == nil {
+            _tableView = UITableView()
+        }
```
這個可以用 `lazy var` 實作，就可以不用判斷 nil 而產物件

Jerry987123 replied:
已修正

markchangjz replied:
ok

markchangjz commented on BootcampTraining/SearchingView/SearchingViewModel/API/iTunesSearchAPIObj.m:
```diff
@@ -0,0 +1,32 @@
+        progress:nil
+         success:^(NSURLSessionDataTask * _Nonnull task, id  _Nullable responseObject) {
+        NSMutableArray <iTunesSearchAPIResponseResult*> *results = [[NSMutableArray alloc] init];
+        int i;
+        for (i=0; i<[[responseObject objectForKey:@"results"] count];i++) {
+            iTunesSearchAPIResponseResult *jsonResponseResult = [[iTunesSearchAPIResponseResult alloc] initWithDictionary: [responseObject objectForKey:@"results"][i] error:nil];
+            if (jsonResponseResult != nil) {
+                [results addObject:jsonResponseResult];
+            }
+        }
```
這段可以用 jsonmodel library 中的 [model-collections](https://github.com/jsonmodel/jsonmodel#model-collections) 修正，就可以不用自己寫 for 迴圈

Jerry987123 replied:
已修正

markchangjz commented on BootcampTraining/SearchingView/SearchingViewController.swift:
```diff
@@ -0,0 +1,71 @@
+        setTableView()
+        setSearchController()
+        setLoadingIndicator()
+        viewModel.movieObservable.subscribe { [weak self] event in
+            if let result = event.element {
+                switch result {
+                case .success(let datas):
+                    if let datas = datas as? [iTunesSearchAPIResponseResult] {
+                        self?.movieDatas = datas
+                        self?._tableView?.reloadData()
+                    }
+                case .failure(let error):
+                    self?.alertWhenAPIError(msg: error.localizedDescription)
+                }
+            }
+        }.disposed(by: disposeBag)
+        viewModel.musicObservable.subscribe { [weak self] event in
+            if let result = event.element {
+                switch result {
+                case .success(let datas):
+                    if let datas = datas as? [iTunesSearchAPIResponseResult] {
+                        self?.musicDatas = datas
+                        self?._tableView?.reloadData()
+                    }
+                case .failure(let error):
+                    self?.alertWhenAPIError(msg: error.localizedDescription)
+                }
+            }
+        }.disposed(by: disposeBag)
```
搜尋一個關鍵字，`self?._tableView?.reloadData()` 會被執行 2 次，可以等到都完成後，再一次呼叫 `reloadData()`，畢竟 `reloadData()` 是有成本的

也可避免網路時間差，先顯示音樂，再插入電影，而造成 UI 閃動

Jerry987123 replied:
已修正

Jerry987123 committed:
unit test中的db insert錯誤處理

markchangjz commented on BootcampTraining/SearchingView/MovieCell.swift:
```diff
@@ -0,0 +1,91 @@
+            DBDao.shared.insertData(mediaType: .movie, model: model)
+            sender.setTitle("取消收藏", for: .normal)
+            collectionButtonWidthConstraint.constant = 80
+        }
```
View 跟 Model 不能無法溝通，不符合 MVC 規則

Cell 不能直接操作 `DBDao.shared.insertData(mediaType: .movie, model: model)`

已知做法，需要用 delegate 或 block 來與 Controller 溝通

<img width="631" alt="截圖 2021-07-02 下午3 18 39" src="https://user-images.githubusercontent.com/15119448/124236526-18cb3c00-db49-11eb-8584-08614b75a716.png">

Jerry987123 replied:
已修正

Jerry987123 committed:
unit test的query方法加入 assert 驗證預期取出幾筆資料

Jerry987123 committed:
[修正] 收藏電影收合狀態錯亂，導致read more或read less顯示錯誤

Jerry987123 committed:
unit test加上assert驗證結果

Jerry987123 committed:
以lzay var改寫tableView初始化

Jerry987123 committed:
以model-collections改寫API response json物件轉換

Jerry987123 committed:
移除音樂收藏錯誤修正

Jerry987123 committed:
tableview reload二次修正成一次

Jerry987123 committed:
修正：不讓Cell直接操作 DBDao

Jerry987123 committed:
DB發生錯誤時，顯示alert

Jerry987123 committed:
修正delegate累積問題

Jerry987123 committed:
dbdao加上對應錯誤處理

markchangjz commented on BootcampTraining/SearchingView/SearchingViewController.swift:
```diff
@@ -0,0 +1,63 @@
+        super.viewWillAppear(animated)
+        tableView.reloadData()
+    }
+    override func viewWillDisappear(_ animated: Bool) {
+        if _searchController?.isActive ?? false {
+            _searchController?.isActive = false
+        }
+    }
```
少呼叫 `super.viewWillDisappear`

Jerry987123 replied:
已修正

markchangjz commented on BootcampTraining/DBTools/DBDaoTest.swift:
```diff
@@ -0,0 +1,59 @@
+//
+//  DBDaoTest.swift
```
預期這個檔案應該會在 `BootcampTrainingTests`資料夾內

Jerry987123 replied:
已修正

Jerry987123 committed:
加上 super.viewWillDisappear

Jerry987123 committed:
將DBDaoTest搬移至BootcampTrainingTests下

markchangjz commented on BootcampTraining/TopicView/TopicViewModel.swift:
```diff
@@ -0,0 +1,25 @@
+}
+
+class TopicViewModel {
+    var datas = ["深色主題", "淺色主題"]
```
data 意為「資料」，是datum的複數，應該是不會有 datas，

建議可以改成 `themes`

Jerry987123 replied:
已修正

markchangjz commented on BootcampTraining/TopicView/TopicViewController.swift:
```diff
@@ -0,0 +1,69 @@
+    }
+}
+extension TopicViewController:UITableViewDataSource {
+    func tableView(_ tableView: UITableView, numberOfRowsInSection section: Int) -> Int {
+        return 2
+    }
```
可以用 `return viewModel.datas.count` 減少 magic number

Jerry987123 replied:
已修正

markchangjz commented on BootcampTraining/TopicView/TopicViewController.swift:
```diff
@@ -0,0 +1,69 @@
+    func tableView(_ tableView: UITableView, cellForRowAt indexPath: IndexPath) -> UITableViewCell {
+        let cell = Bundle.main.loadNibNamed("TopicCell", owner: self, options: nil)?.first as! TopicCell
+        cell.titleLabel.text = viewModel.datas[indexPath.row]
+        if viewModel.getSelectedTopic().rawValue == indexPath.row {
+            cell.accessoryType = .checkmark
+        }
```
記得加 else `cell.accessoryType = .none`，不然當資料量多的時候，table view 上下卷到 reuse cell 會冒出很多 checkmark

Jerry987123 replied:
已修正

markchangjz commented on BootcampTraining/TopicView/TopicViewController.swift:
```diff
@@ -0,0 +1,69 @@
+    }
+}
+extension TopicViewController:UITableViewDelegate {
+    func tableView(_ tableView: UITableView, didSelectRowAt indexPath: IndexPath) {
+        switch indexPath.row {
+        case 0:
+            viewModel.setSelectedTopic(topicType: .deepColor)
+        case 1:
+            viewModel.setSelectedTopic(topicType: .lightColor)
+        default:
+            break
+        }
+        tableView.reloadData()
+        if let tabBar = tabBarController as? MainTabBarViewController {
+            TopicInteractor.readTopicListFile()
+            tabBar.tabbarTopicColorSet()
+        }
+        setUI()
+    }
```
每選完一次 `tableView.reloadData()`，建議只刷新取消選取 cell 跟被點選的 cell

可參考這樣的做法
```
- (void)tableView:(UITableView *)tableView didSelectRowAtIndexPath:(NSIndexPath *)indexPath {
	MKCTheme currentTheme = MKCThemeManager.currentTheme;
	
	NSIndexPath *currentSelectedIndexPath = [NSIndexPath indexPathForRow:currentTheme inSection:0];
	
	MKCTheme applyTheme = (MKCTheme)indexPath.row;
	[MKCThemeManager applyTheme:applyTheme];
	
	[tableView reloadRowsAtIndexPaths:@[currentSelectedIndexPath] withRowAnimation:UITableViewRowAnimationFade];
	
	UITableViewCell *cell = [tableView cellForRowAtIndexPath:indexPath];
	cell.accessoryType = UITableViewCellAccessoryCheckmark;
	[tableView deselectRowAtIndexPath:indexPath animated:YES];
}
```

Jerry987123 replied:
已修正

Jerry987123 committed:
datas改成themes、return 2改成return ViewModel.themes.count

Jerry987123 committed:
修正accessoryType的 if else 給值增加else項為 .none

markchangjz commented on BootcampTraining/CollectionView/CollectionViewControllerDelegate.swift:
```diff
@@ -0,0 +1,117 @@
+            let cell = Bundle.main.loadNibNamed("MovieCell", owner: self, options: nil)?.first as! MovieCell
+            cell.setCell(model: movieTableDatas[indexPath.row])
+            cell.expandCell = { [weak self] sender in
+                guard let cell = sender.superview?.superview as? MovieCell else {
```
`guard let cell = sender.superview?.superview as? MovieCell else {`

這種寫法，只要有人動 MovieCell.xib 階層就很容易導致改 A 壞 B，要再想看看有沒有比較好的做法

Jerry987123 replied:
已修正

markchangjz commented on BootcampTraining/CollectionView/CollectionViewControllerDelegate.swift:
```diff
@@ -0,0 +1,117 @@
+    func tableView(_ tableView: UITableView, cellForRowAt indexPath: IndexPath) -> UITableViewCell {
+        switch switchButtonView.selectedSegmentIndex {
+        case 0:
+            let cell = Bundle.main.loadNibNamed("MovieCell", owner: self, options: nil)?.first as! MovieCell
```
`let cell = Bundle.main.loadNibNamed("MovieCell", owner: self, options: nil)?.first as! MovieCell`

沒用到 `dequeueReusableCell` 感覺會很吃記憶體用量

Jerry987123 replied:
已修正

Jerry987123 committed:
設定主題方法名稱調整

Jerry987123 committed:
只刷新須取消點選的cell和被選點的Cell，取代tableView.reloadData()

markchangjz commented on BootcampTraining/SearchingView/MovieCell.swift:
```diff
@@ -0,0 +1,101 @@
+            photoImageView.image = UIImage.init(systemName: "tv")
+        }
+        if let trackId = model.trackId {
+            let alreadyAdded = CollectionInteractor().alreadyAdded(trackId: Int(truncating: trackId))
```
`let alreadyAdded = CollectionInteractor().alreadyAdded(trackId: Int(truncating: trackId))`

* 在 View 中直接存取 DB model 不符合 MVC

Jerry987123 replied:
已修正

Jerry987123 committed:
Bundle.main.loadNibName改用dequeueReusableCell

Jerry987123 committed:
track time無值時顯示空字串

markchangjz commented on BootcampTraining/SearchingView/MusicCell.swift:
```diff
@@ -0,0 +1,78 @@
+        artiestNameLabel.text = model.artistName
+        collectionNameLabel.text = model.collectionName
+        musicTimeLabel.text = timeFromMillisToHMMSS(time: Int(truncating: model.trackTimeMillis ?? 0))
+        if let artworkUrl100 = model.artworkUrl100, let url = URL(string: artworkUrl100), UIApplication.shared.canOpenURL(url) {
+            photoImageView.sd_setImage(with: url)
+        } else {
+            photoImageView.image = UIImage.init(systemName: "music.note")
+        }
```
可以用 `photoImageView.sd_setImage(with: url, placeholderImage: UIImage.init(systemName: "music.note"))` 取代

Jerry987123 replied:
已修正

markchangjz commented on BootcampTraining/AboutiTunes/AboutiTunesViewController.swift:
```diff
@@ -0,0 +1,45 @@
+        loadWebView()
+    }
+    private func setWebView(){
+        if _webView == nil {
+            _webView = WKWebView()
+        }
+        guard let webView = _webView else {
+            return print("webview failed to init")
+        }
```
用 `lazy var`

Jerry987123 replied:
已修正

Jerry987123 committed:
webView初始化改用lazy var

Jerry987123 committed:
圖片下載出錯時，改顯示預設圖
UIImageView.sd_setImage(with:url) 改成 UIImageView.sd_setImage(with:url, placeholderImage: UIImage(預設圖))

Jerry987123 committed:
Cell中不直接呼叫CollectionInteractor()

Jerry987123 committed:
將sender.superview?.superview寫法改成以trackid為唯一值做尋找資料排序位置

markchangjz commented on BootcampTrainingTests/DBDaoTest.swift:
```diff
@@ -0,0 +1,61 @@
+        case .success:
+            assert(true)
+        case .failure(let error):
+            assert(false, "delete未成功, error=\(error.localizedDescription)")
```
`assert(false, "delete未成功, error=\(error.localizedDescription)")`

要使用 XCTAssert，有錯誤時才會寫 fail

Jerry987123 replied:
已修正

markchangjz commented on BootcampTrainingTests/BootcampTrainingTests.swift:
```diff
@@ -0,0 +1,24 @@
+        print("--------------------")
+    }
+
+    func testExample() throws {
```
`func testExample() throws {`

一般來說一個案例只會測一種功能，預期來說會分開測試 查詢、刪除、新增

Jerry987123 replied:
已修正

markchangjz commented on BootcampTraining/SearchingView/SearchingViewController.swift:
```diff
@@ -0,0 +1,64 @@
+class SearchingViewController: UIViewController {
+
+    lazy var tableView = UITableView()
+    var _searchController:UISearchController?
+    var _loadingIndicator:UIActivityIndicatorView?
```
lazy var

Jerry987123 replied:
已修正

markchangjz commented on BootcampTraining/SearchingView/MovieCell.swift:
```diff
@@ -0,0 +1,101 @@
+        }
+    }
+    private func timeFromMillisToHMMSS(time:Int) -> String {
+        let hour = time/1000/3600
+        let minMinusHour = time/1000%3600
+        let min = minMinusHour/60
+        let sec = minMinusHour%60
+        let tools = Tools()
+        return "\(hour):\(tools.make0To00(number: min)):\(tools.make0To00(number: sec))"
```
可以用 [NSDateComponentsFormatter](https://crunchybagel.com/formatting-a-duration-with-nsdatecomponentsformatter/) 內建 method 優化

Jerry987123 replied:
已修正

markchangjz commented on BootcampTraining/Tools/Tools.swift:
```diff
@@ -0,0 +1,27 @@
+            return numberString
+        }
+    }
+    func numberCurrency(count:Int) -> String {
```
`func numberCurrency(count:Int) -> String {`

這個其實可以不用 instance method，感覺用 class method 做比較適合

Jerry987123 replied:
已修正

markchangjz commented on BootcampTraining/CollectionView/CollectionViewController.swift:
```diff
@@ -0,0 +1,82 @@
+            break
+        }
+    }
+    @IBAction func switchButtonAction(_ sender: UISegmentedControl) {
+        switch sender.selectedSegmentIndex {
+        case 0:
+            viewModel.loadCollectionFromDB(mediaType: .movie)
+        case 1:
+            viewModel.loadCollectionFromDB(mediaType: .music)
+        default:
+            break
+        }
+        tableView.reloadData()
+    }
```
收藏頁面的 table view 都共用同一個，當電影類往上滑 100，切到音樂就會跟著往上滑 100，體驗比較差

Jerry987123 replied:
已修正

Jerry987123 committed:
將未初始化的變數以lazy var設定

Jerry987123 committed:
將千位數func移出Tools

Jerry987123 committed:
unit test改成一個案例測一種功能、assert修正成XCTAssert

Jerry987123 committed:
被觀察者Behavior改成Publish

Jerry987123 committed:
新增APITest

Jerry987123 committed:
名稱調整、頁面調整、移除多餘程式碼

Jerry987123 committed:
用DateComponentsFormatter修正時間格式

Jerry987123 committed:
多一個大括號移除

Jerry987123 committed:
收藏頁面電影和音樂分屬不同tableView，增加滑動體驗
